src/codec.rs

```diff
@@ -619,16 +626,47 @@ impl<T: Decode> Decode for Vec<T> {
 		<Compact<u32>>::decode(input).and_then(move |Compact(len)| {
 			let len = len as usize;
 			if let IsU8::Yes = <T as Decode>::IS_U8 {
-				if len > input.remaining_len()? {
-					return Err("Not enough data to decode vector".into());
+
+				let input_len = input.remaining_len()?;
+
+				// If there is input len and it cannot be pre-allocated then return directly.
+				if input_len.map(|l| l < len).unwrap_or(false) {
+					return Err("Not enough data to decode vector".into())
 				}
 
-				let mut r = vec![0; len];
-				input.read(&mut r)?;
+				// Note: we checked that if input_len is some then it can preallocated.
+				let r = if input_len.is_some() || len < MAX_PREALLOCATION {
+
+					// Here we pre-allocate the whole buffer.
+					let mut r = vec![0; len];
+					input.read(&mut r)?;
+
+					r
+				} else {
+
+					// Here we pre-allocate only the maximum pre-allocation
+					let mut r = vec![];
+
+					let mut remains = len;
+					let buffer_len = MAX_PREALLOCATION;
+					let mut buffer = vec![0; buffer_len];
+
+					while remains != 0 {
+						let len_read = buffer_len.min(remains);
+						input.read(&mut buffer[..len_read])?;
+						r.extend_from_slice(&buffer[..len_read]);
```

Comment:
If you would use `resize`, you would not need `buffer`.

Reply:
done